yt/data_objects/construction_data_containers.py

```diff
@@ -240,12 +251,32 @@ def field(self):
 
     def get_data(self, fields=None):
         fields = self._determine_fields(fields)
+        sfields = []
+        if self.moment == 2:
+
+            def make_sq_field(fname):
+                def _sq_field(field, data):
+                    return data[fname] ** 2
+
+                return _sq_field
+
+            for field in fields:
+                fd = self.ds._get_field_info(*field)
+                self.ds.add_field(
+                    (field[0], f"tmp_{field[1]}_squared"),
+                    make_sq_field(field),
```

Comment:
This is fragile, using `functools.partial` for avoid late-binding issues, see https://github.com/yt-project/yt/pull/3996